# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [3]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline
pd.options.mode.chained_assignment = None
from sklearn.cluster import KMeans

In [4]:
cd

C:\Users\sjrek


In [5]:
# correct the path
path="C:\Springboard\Github\Guided Capstone"
os.chdir(path) 

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [6]:
ski = pd.read_csv("step3_output.csv")
ski.head(5)

Unnamed: 0                           Name       state  summit_elev  \
0           2               Hilltop Ski Area      Alaska         2090   
1           4            Sunrise Park Resort     Arizona        11100   
2           5  Yosemite Ski & Snowboard Area  California         7800   
3           8         Boreal Mountain Resort  California         7700   
4           9                    Dodge Ridge  California         8200   

   vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  ...  \
0            294      0        0.0          0          0     0  ...   
1           1800      0        0.0          0          1     2  ...   
2            600      0        0.0          0          0     0  ...   
3            500      0        0.0          0          1     1  ...   
4           1600      0        0.0          0          0     1  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0            30.0        150.000000       36.0   
1              800.0            80.0        115.000000       49.0   
2               88.0             0.0        110.000000       84.0   
3              380.0           200.0        150.000000       54.0   
4              862.0             0.0        115.103943       69.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0          30.0          34.0              152.0   
1            250.0          74.0          78.0              104.0   
2            300.0          47.0          47.0              107.0   
3            400.0          49.0          49.0              150.0   
4            350.0          78.0          78.0              140.0   

   NightSkiing_ac  clusters  
0            30.0         2  
1            80.0         1  
2             0.0         1  
3           200.0         1  
4             0.0         1  

[5 rows x 27 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [7]:
# I've not created dummy variables before, so I will break this down further
dfo=ski['state'] # select states
dummy = pd.get_dummies(dfo)
ski = pd.concat([ski.drop('state', axis=1), dummy], axis=1)

In [8]:
ski.head()

Unnamed: 0                           Name  summit_elev  vertical_drop  \
0           2               Hilltop Ski Area         2090            294   
1           4            Sunrise Park Resort        11100           1800   
2           5  Yosemite Ski & Snowboard Area         7800            600   
3           8         Boreal Mountain Resort         7700            500   
4           9                    Dodge Ridge         8200           1600   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  Rhode Island  \
0      0        0.0          0          0     0       1  ...             0   
1      0        0.0          0          1     2       3  ...             0   
2      0        0.0          0          0     0       1  ...             0   
3      0        0.0          0          1     1       3  ...             0   
4      0        0.0          0          0     1       2  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 61 columns]

"Unnamed: 0" column does not belong. I'm going to drop it.

In [32]:
ski.drop("Unnamed: 0", axis=1, inplace=True)

In [52]:
ski.describe().transpose()

count         mean          std    min      25%     50%  \
summit_elev        172.0  3637.656977  3223.022158  315.0  1245.00  2000.0   
vertical_drop      172.0   841.447674   584.594889  175.0   391.25   650.0   
trams              172.0     0.000000     0.000000    0.0     0.00     0.0   
fastEight          172.0     0.000000     0.000000    0.0     0.00     0.0   
fastSixes          172.0     0.000000     0.000000    0.0     0.00     0.0   
fastQuads          172.0     0.238372     0.588520    0.0     0.00     0.0   
quad               172.0     0.540698     0.703854    0.0     0.00     0.0   
triple             172.0     1.209302     1.234175    0.0     0.00     1.0   
double             172.0     1.720930     1.472168    0.0     1.00     1.0   
surface            172.0     2.261628     1.445353    0.0     1.00     2.0   
total_chairs       172.0     5.970930     2.768721    0.0     4.00     5.0   
Runs               172.0    29.715116    17.443148    7.0    16.00    24.0   
TerrainParks       172.0     1.761628     1.412614    0.0     1.00     2.0   
LongestRun_mi      172.0     0.989535     0.741585    0.0     0.40     1.0   
SkiableTerrain_ac  172.0   230.034884   287.013293    8.0    50.00   117.5   
Snow Making_ac     172.0    82.005814    71.350096    0.0    30.00    60.0   
daysOpenLastYear   172.0   106.146641    22.026385   56.0    94.75   110.0   
yearsOpen          172.0    59.441860    14.247277   22.0    51.75    58.0   
averageSnowfall    172.0   145.142920   106.155149   18.0    51.75   120.0   
AdultWeekday       172.0    50.108117    16.070413   17.0    39.00    49.0   
AdultWeekend       172.0    56.397543    15.234606   20.0    45.00    55.0   
projectedDaysOpen  172.0   112.280652    21.200225   56.0   100.00   115.0   
NightSkiing_ac     172.0    44.296512    47.442787    0.0     0.00    35.0   
clusters           172.0     1.430233     0.749904    0.0     1.00     2.0   
Alaska             172.0     0.005814     0.076249    0.0     0.00     0.0   
Arizona            172.0     0.005814     0.076249    0.0     0.00     0.0   
California         172.0     0.069767     0.255498    0.0     0.00     0.0   
Colorado           172.0     0.023256     0.151155    0.0     0.00     0.0   
Connecticut        172.0     0.029070     0.168493    0.0     0.00     0.0   
Idaho              172.0     0.034884     0.184021    0.0     0.00     0.0   
Illinois           172.0     0.017442     0.131293    0.0     0.00     0.0   
Indiana            172.0     0.011628     0.107517    0.0     0.00     0.0   
Iowa               172.0     0.017442     0.131293    0.0     0.00     0.0   
Maine              172.0     0.029070     0.168493    0.0     0.00     0.0   
Maryland           172.0     0.005814     0.076249    0.0     0.00     0.0   
Massachusetts      172.0     0.023256     0.151155    0.0     0.00     0.0   
Michigan           172.0     0.087209     0.282965    0.0     0.00     0.0   
Minnesota          172.0     0.046512     0.211205    0.0     0.00     0.0   
Missouri           172.0     0.011628     0.107517    0.0     0.00     0.0   
Montana            172.0     0.017442     0.131293    0.0     0.00     0.0   
Nevada             172.0     0.005814     0.076249    0.0     0.00     0.0   
New Hampshire      172.0     0.063953     0.245384    0.0     0.00     0.0   
New Jersey         172.0     0.005814     0.076249    0.0     0.00     0.0   
New Mexico         172.0     0.034884     0.184021    0.0     0.00     0.0   
New York           172.0     0.116279     0.321495    0.0     0.00     0.0   
North Carolina     172.0     0.017442     0.131293    0.0     0.00     0.0   
Ohio               172.0     0.023256     0.151155    0.0     0.00     0.0   
Oregon             172.0     0.017442     0.131293    0.0     0.00     0.0   
Pennsylvania       172.0     0.063953     0.245384    0.0     0.00     0.0   
Rhode Island       172.0     0.005814     0.076249    0.0     0.00     0.0   
South Dakota       172.0     0.005814    

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [33]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = ski.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = ski.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [34]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

Before using these sets to create a model, I want to look at them to make sure they make sense

In [35]:
y_test.mean()

57.258533872598576

In [36]:
y_train.mean()

56.110546006066734

In [37]:
X_train

array([[-0.43180235,  0.52934642,  0.        , ..., -0.13323468,
        -0.248452  , -0.17303214],
       [-0.71652628, -0.58578104,  0.        , ..., -0.13323468,
        -0.248452  , -0.17303214],
       [-0.60761549, -0.64582637,  0.        , ..., -0.13323468,
        -0.248452  , -0.17303214],
       ...,
       [ 1.51925672,  2.02190163,  0.        , ..., -0.13323468,
        -0.248452  , -0.17303214],
       [ 0.85956849,  0.20338608,  0.        , ..., -0.13323468,
        -0.248452  , -0.17303214],
       [-0.79587557,  0.0146722 ,  0.        , ..., -0.13323468,
        -0.248452  , -0.17303214]])

In [50]:
X_train.max()

13.076696830622025

This number is way too high and would be an outlier. What happened? I need to go back and look at my scaling again.

In [51]:
X_test.max()

13.076696830622023

#### Model 1

In [46]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [47]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [48]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
from sklearn.metrics import mean_squared_error
# You can plug y_test and y_pred into the functions to evaluate the model
m1_evs = explained_variance_score(y_test, y_pred)
m1_evs

-1.962081174491013e+21

In [49]:
m1_mae = mean_absolute_error(y_test, y_pred)
m1_mae

168200410182.6963

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [42]:
print(lm.intercept_)

42050102600.75715


This Value is WAY off the mean Adult Weekend. There has to be some issue with this model. I will check the predicted values to see how that turned out.

In [43]:
y_pred.mean()

168200410234.68936

Predicted values also have a mean value that doesn't make any sense. I need to review this model to find out what's wrong.

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [60]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coff_m1 = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
print(coff_m1.sort_values('Coefficient', ascending=False))

                    Coefficient
trams              2.494771e+13
fastEight          1.267117e+13
fastSixes          8.457336e+12
total_chairs       7.685073e+12
double             4.086261e+12
surface            4.011832e+12
triple             3.425670e+12
quad               1.953670e+12
fastQuads          1.633539e+12
New York           1.159272e+12
Michigan           1.020338e+12
California         9.212961e+11
Pennsylvania       8.848261e+11
New Hampshire      8.848261e+11
Wisconsin          8.462645e+11
Minnesota          7.615800e+11
Vermont            7.615800e+11
Idaho              6.635571e+11
New Mexico         6.635571e+11
Wyoming            6.075638e+11
Connecticut        6.075638e+11
Maine              6.075638e+11
Ohio               5.450461e+11
Massachusetts      5.450461e+11
Colorado           5.450461e+11
Illinois           4.734265e+11
Montana            4.734265e+11
Iowa               4.734265e+11
Oregon             4.734265e+11
Utah               4.734265e+11
Washingt

This model is so ridiculously off, that I can't look at these values as in a reasonable way. Instead, I'm going to redo the model using random forest to see if the result are any better before moving on to Model 2.

In [64]:
from sklearn.ensemble import RandomForestRegressor
regressor = RandomForestRegressor(random_state=0, n_estimators=200)
model1 = regressor.fit(X_train, y_train)
model1.score(X_test, y_test)

0.7973120214187768

This seems like a much better model. Let me look at the same tests what was run on the other model.

In [70]:
y_pred1 = model1.predict(X_test)

In [71]:
m1_evs = explained_variance_score(y_test, y_pred1)
m1_evs

0.8102400964530954

In [72]:
m1_mae = mean_absolute_error(y_test, y_pred1)
m1_mae

6.170408847320527

These look way better than what we got from the previous model. Now, let's take a look at the intercept and coefficients.

In [76]:
#model1.feature_importances_
coff_m1 = pd.DataFrame(model1.feature_importances_, X.columns, columns=['Importance'])
print(coff_m1.sort_values('Importance', ascending=False))

                   Importance
AdultWeekday         0.664255
Snow Making_ac       0.066554
vertical_drop        0.027955
averageSnowfall      0.021874
LongestRun_mi        0.021764
Runs                 0.019431
summit_elev          0.019369
SkiableTerrain_ac    0.018271
total_chairs         0.017111
NightSkiing_ac       0.016767
yearsOpen            0.015815
projectedDaysOpen    0.012245
clusters             0.011013
North Carolina       0.008582
triple               0.008191
daysOpenLastYear     0.006541
TerrainParks         0.006449
double               0.005310
quad                 0.005088
Vermont              0.004478
surface              0.004262
Virginia             0.002577
New Hampshire        0.002403
New York             0.002164
Maine                0.001560
Michigan             0.001408
Pennsylvania         0.001125
Connecticut          0.001081
Idaho                0.000925
Wisconsin            0.000923
Tennessee            0.000741
West Virginia        0.000715
fastQuads 

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

I wound up with very different results since I ran the random forest fit instead. The Adult Weekday was obviously closely related to the Adult Weekend but is a response variable that we are looking to modify as well. We should not include this in the model. I will make that adjustment and run it again for Model 2.

#### Model 2

In [77]:
X = ski.drop(['Name','AdultWeekend', 'AdultWeekday'], axis=1)

In [78]:
# new values of X, so it needs to be scaled again
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 

In [80]:
# y variable should remain the same, so we can use that again, but need different samples
# Recreate the four variables: X_train, X_test, y_train and y_test
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [81]:
# create the next model
regressor = RandomForestRegressor(random_state=0, n_estimators=200)
model2 = regressor.fit(X_train, y_train)
model2.score(X_test, y_test)

0.44404603489004113

This is surprisingly a better score without considering the Adult Weekday. Let's see the other scores.

In [82]:
y_pred2 = model2.predict(X_test)

In [83]:
m2_evs = explained_variance_score(y_test, y_pred2)
m2_evs

0.44979314007964954

In [84]:
m2_mae = mean_absolute_error(y_test, y_pred2)
m2_mae

9.854552072800807

Explained Variance is smaller but Mean Absolute Error is larger. Either way, we can't really consider Adult Weekday ticket prices as factor for increasing the ticket price at our resort, so I would highly favor leaving this out of the model.
Let's look at the feature importances again to see if something else should be eliminated.

In [85]:
#model2.feature_importances_
coff_m2 = pd.DataFrame(model2.feature_importances_, X.columns, columns=['Importance'])
print(coff_m2.sort_values('Importance', ascending=False))

                   Importance
Snow Making_ac       0.239645
Runs                 0.203795
vertical_drop        0.114670
LongestRun_mi        0.084628
clusters             0.036292
summit_elev          0.035544
total_chairs         0.032016
yearsOpen            0.031267
daysOpenLastYear     0.027993
projectedDaysOpen    0.027278
SkiableTerrain_ac    0.025470
averageSnowfall      0.019961
NightSkiing_ac       0.019033
triple               0.015597
surface              0.010897
Vermont              0.009993
quad                 0.009439
double               0.008450
fastQuads            0.007762
North Carolina       0.007082
TerrainParks         0.006523
New Hampshire        0.004904
Virginia             0.003889
New York             0.003513
West Virginia        0.002644
Michigan             0.001722
California           0.001614
Wisconsin            0.001311
Pennsylvania         0.001255
Rhode Island         0.000709
Connecticut          0.000599
Idaho                0.000555
Minnesota 

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

I come to very different conclusion. Snowmaking looks like it takes top priority in this model. This probably is is more prominent in the smaller mountains that make up a larger part of our sample. Since there are a lot of small mountains in NY state and New England, this is very different than Northwest Montana. We should eliminate snowmaking from the model.

#### Model 3

In [86]:
X = ski.drop(['Name','AdultWeekend', 'AdultWeekday', 'Snow Making_ac'], axis=1)

In [87]:
# new values of X, so it needs to be scaled again
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 

In [88]:
# y variable should remain the same, so we can use that again, but need different samples
# Recreate the four variables: X_train, X_test, y_train and y_test
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [89]:
# create the next model
regressor = RandomForestRegressor(random_state=0, n_estimators=200)
model3 = regressor.fit(X_train, y_train)
model3.score(X_test, y_test)

0.430407664554544

This is a slightly better score than the previous 2 models. Let's see how the other scores look.

In [90]:
y_pred3 = model3.predict(X_test)

In [91]:
m3_evs = explained_variance_score(y_test, y_pred3)
m3_evs

0.4377998413405766

In [92]:
m3_mae = mean_absolute_error(y_test, y_pred3)
m3_mae

10.311052325581393

Just as before, the explained variance went down but the mean absolute error increased. Model 2 maybe a better selection. To be complete in this modeling, I want to run the models with the values removed that I was instructed to remove; I will do 2 more runs with this.

#### Model 4

In this approach, I will eliminate the state variables as we were instructed above. I will still use the random forest approach.

In [94]:
# determine the columns to drop
ski.columns

Index(['Name', 'summit_elev', 'vertical_drop', 'trams', 'fastEight',
       'fastSixes', 'fastQuads', 'quad', 'triple', 'double', 'surface',
       'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac', 'clusters', 'Alaska', 'Arizona', 'California',
       'Colorado', 'Connecticut', 'Idaho', 'Illinois', 'Indiana', 'Iowa',
       'Maine', 'Maryland', 'Massachusetts', 'Michigan', 'Minnesota',
       'Missouri', 'Montana', 'Nevada', 'New Hampshire', 'New Jersey',
       'New Mexico', 'New York', 'North Carolina', 'Ohio', 'Oregon',
       'Pennsylvania', 'Rhode Island', 'South Dakota', 'Tennessee', 'Utah',
       'Vermont', 'Virginia', 'Washington', 'West Virginia', 'Wisconsin',
       'Wyoming'],
      dtype='object')

In [96]:
state_list = ['Alaska', 'Arizona', 'California',
       'Colorado', 'Connecticut', 'Idaho', 'Illinois', 'Indiana', 'Iowa',
       'Maine', 'Maryland', 'Massachusetts', 'Michigan', 'Minnesota',
       'Missouri', 'Montana', 'Nevada', 'New Hampshire', 'New Jersey',
       'New Mexico', 'New York', 'North Carolina', 'Ohio', 'Oregon',
       'Pennsylvania', 'Rhode Island', 'South Dakota', 'Tennessee', 'Utah',
       'Vermont', 'Virginia', 'Washington', 'West Virginia', 'Wisconsin',
       'Wyoming']
ski.drop(state_list, axis=1, inplace=True)

In [97]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = ski.drop(['Name','AdultWeekend'], axis=1)

In [98]:
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [99]:
# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [100]:
# create the next model
regressor = RandomForestRegressor(random_state=0, n_estimators=200)
model4 = regressor.fit(X_train, y_train)
model4.score(X_test, y_test)

0.7877018421204228

This model has a worse score; let's see how it does on the other tests.

In [101]:
y_pred4 = model4.predict(X_test)

In [102]:
m4_evs = explained_variance_score(y_test, y_pred4)
m4_evs

0.8039159877302895

In [103]:
m4_mae = mean_absolute_error(y_test, y_pred4)
m4_mae

6.353073053589486

These scores are better than the result from model 2 or 3; this is actually a better model to exclude the states. Let's look at the importance of the factors.

In [104]:
#model3.feature_importances_
coff_m4 = pd.DataFrame(model4.feature_importances_, X.columns, columns=['Importance'])
print(coff_m4.sort_values('Importance', ascending=False))

                   Importance
AdultWeekday         0.665483
Snow Making_ac       0.068638
vertical_drop        0.029270
averageSnowfall      0.024280
LongestRun_mi        0.023003
Runs                 0.021639
summit_elev          0.021632
yearsOpen            0.020650
SkiableTerrain_ac    0.020250
total_chairs         0.017463
NightSkiing_ac       0.016971
clusters             0.015536
projectedDaysOpen    0.013232
triple               0.009156
daysOpenLastYear     0.009039
TerrainParks         0.007049
double               0.005961
quad                 0.005151
surface              0.004826
fastQuads            0.000771
fastSixes            0.000000
fastEight            0.000000
trams                0.000000


That makes sense that if the model includes Adult Weekday that it would be more accurate in predicting the Weekend price. This is so closely correlated that it seems like cheating, as this variable is something we are trying to determine for our own Big Mountain Resort. I don't believe this is a good approach other than to realize that the the price of weekend ticket is very closely related to the weekend ticket. This is the model with 'state' removed; now, I will follow instruction and also remove summit elevation (base elevation was already removed back in step 3).

#### Model 5

In [105]:
# drop columns that we don't want
X = ski.drop(['Name','AdultWeekend', 'summit_elev'], axis=1)
X.columns

Index(['vertical_drop', 'trams', 'fastEight', 'fastSixes', 'fastQuads', 'quad',
       'triple', 'double', 'surface', 'total_chairs', 'Runs', 'TerrainParks',
       'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac',
       'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday',
       'projectedDaysOpen', 'NightSkiing_ac', 'clusters'],
      dtype='object')

These look right. Now to make the split again.

In [106]:
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [107]:
# create the next model
regressor = RandomForestRegressor(random_state=0, n_estimators=200)
model5 = regressor.fit(X_train, y_train)
model5.score(X_test, y_test)

0.7939868280905577

In [109]:
y_pred5 = model5.predict(X_test)

In [110]:
explained_variance_score(y_test, y_pred5)

0.8096992535606947

In [111]:
mean_absolute_error(y_test, y_pred5)

6.297602578361983

These numbers are slightly better than the previous model. Perhaps we drop summit elevation.

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.810 | 6.17 |-|
| Model 2. | 0.450 | 9.85 |'AdultWeekday'|
| Model 3. | 0.438 | 10.31 |'AdultWeekday', 'Snow Making_ac'|
| Model 4. | 0.804 | 6.35 |'state'|
| Model 5. | 0.810 | 6.30 |'state', 'summit_elev'|

#### Model Selection:

Model 1 and Model 5 appears to provide the best results, but we should not be including Adult Weekday price in the prediction of Adult Weekend price. Both of these are variables of interest that we will be looking to modify. The instructions stated "Later we will go back and consider the AdultWeekday, dayOpenLastYear, and projectedDaysOpen". I didn't see this happen in this Step of the process. 

Without the context of the problem, I would select Model 5 - it has great error results that are very similar to Model 1, but without the extra variables. However, I wouldn't feel confident in any of these models to predict the lift ticket price increase for Big Mountain Resort. We had a data set that Big Mountain would have fallen inside of, letting us do an interpolation; rather than throw out some of the smaller mountains that are no similar to Big Mountain, we threw out the bigger mountains that were outliers for skiable acres and vertical. Therefore we would be extrapolating on these variables when we didn't have to. We should have thrown out the small mountains because they are not similar to what we were looking for.

We were considering adjusting the lift ticket prices for weekends and weekdays; they are closely related at ski resorts for obvious reasons. To include one in the model of of the other doesn't seem like a wise choice, as it would overpower all of the other variables. 